internal/writeaheadlog/message_wal.go

```diff
@@ -0,0 +1,54 @@
+package writeaheadlog
+
+import (
+	"github.com/filecoin-project/go-f3/gpbft"
+)
+
+type MessageWriteAheadLog struct {
+	delegate *WriteAheadLog[MessageEntry, *MessageEntry]
+}
+
+func OpenMessageWriteAheadLog(path string) (*MessageWriteAheadLog, error) {
+	delegate, err := Open[MessageEntry, *MessageEntry](path)
+	if err != nil {
+		return nil, err
+	}
+	return &MessageWriteAheadLog{
+		delegate: delegate,
+	}, nil
+}
+
+func (mwal *MessageWriteAheadLog) ForEach(apply func(message *gpbft.GMessage) bool) error {
+	return mwal.delegate.ForEach(
+		func(entry MessageEntry) bool {
+			return apply(entry.Message)
+		},
+	)
+}
+
+func (mwal *MessageWriteAheadLog) Append(msg *gpbft.GMessage) error {
+	return mwal.delegate.Append(MessageEntry{Message: msg})
+}
+
+func (mwal *MessageWriteAheadLog) Purge(upToEpoch uint64) error {
+	return mwal.delegate.Purge(upToEpoch)
+}
+
+func (mwal *MessageWriteAheadLog) FindMessages(instance, round uint64, phase gpbft.Phase) ([]*gpbft.GMessage, error) {
+	var matches []*gpbft.GMessage
+	// Return partial matches even if error is non-nil. This is better than "all or
+	// nothing" since it offers more information and the GPBFT is meant to tolerate
+	// partial delivery anyway.
+	return matches, mwal.ForEach(
+		func(subject *gpbft.GMessage) bool {
+			if subject.Vote.Instance == instance && subject.Vote.Round == round && subject.Vote.Phase == phase {
+				matches = append(matches, subject)
+			}
+			return true
+		},
+	)
+}
```

Comment:
This will re-read from disk every time we rebroadcast? That's not great.

Can we not create an index on start?

Reply:
Are we OK keeping the entire WAL in memory?

Reply:
At least for a fixed number of instances, yes. As @Kubuxu pointed out last time we talked, given the exponential backoff nature of rounds, we'll never have to remember that many messages.

Reply:
(I'd prefer to just remember one instance, but whatever is easiest).

Reply:
The limit of 5 instances in WAL is externally enforced, via runner start loop. If for whatever reason logs are partially removed the hydration will load all the logs.

I am going to move that cap into WAL itself for safety.

Reply:
This is going to lead to a whole bunch of unnecessary code churn, IMO. From what I can tell, all the WAL changes were introduced primarily to support `ForEach`, which isn't necessary in this case.

Reply:
OK
Just so I get it right: instead of `forEach`, you'd like to see call to `All`, then filter. Correct?

Reply:
Discussed on slack.